## Hacker News Submissions Project

In this project, we'll work with a dataset of submissions to popular technology site Hacker News.

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") receive votes and comments, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of the Hacker News listings can get hundreds of thousands of visitors as a result.

we have reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that didn't receive any comments and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

* id: the unique identifier from Hacker News for the post
* title: the title of the post
* url: the URL that the posts links to, if the post has a URL
* num_points: the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* num_comments: the number of comments on the post
* author: the username of the person who submitted the post
* created_at: the date and time of the post's submission


<b> We're specifically interested in posts with titles that begin with either Ask HN or Show HN. </b>

Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a few examples:

* Ask HN: How to improve my personal website?
* Ask HN: Am I the only one outraged by Twitter shutting down share counts?
* Ask HN: Aby recent changes to CSS that broke mobile?

* Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just something interesting. Below are a few examples:

* Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
* Show HN: Something pointless I made
* Show HN: Shanhu.io, a programming playground powered by e8vm

<b>We'll compare these two types of posts to determine the following:</b>

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?

In [2]:
# reading the hacker_news file
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
hn_columns = hn[0]
hn =hn[1:]

In [3]:
#Displaying Headers
print(hn_columns)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [4]:
#displaying first 5 rows
print(hn[0:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

To find the posts that begin with either Ask HN or Show HN, we'll use the string method startswith.

In [5]:
ask_posts = []
show_posts = []
other_posts = []

In [6]:
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [7]:
#length of various categories
print('ask_posts', len(ask_posts))
print('show_posts', len(show_posts))
print('other_posts', len(other_posts))

ask_posts 1744
show_posts 1162
other_posts 17194


In [8]:
# First five rows of ask posts
ask_posts[:5]

[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43'],
 ['11610310',
  'Ask HN: Aby recent changes to CSS that broke mobile?',
  '',
  '1',
  '1',
  'polskibus',
  '5/2/2016 10:14'],
 ['12210105',
  'Ask HN: Looking for Employee #3 How do I do it?',
  '',
  '1',
  '3',
  'sph130',
  '8/2/2016 14:20'],
 ['10394168',
  'Ask HN: Someone offered to buy my browser extension from me. What now?',
  '',
  '28',
  '17',
  'roykolak',
  '10/15/2015 16:38']]

In [9]:
#first 5 rows of show_posts
show_posts[:5]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46'],
 ['11590768',
  'Show HN: Shanhu.io, a programming playground powered by e8vm',
  'https://shanhu.io',
  '1',
  '1',
  'h8liu',
  '4/28/2016 18:05'],
 ['12178806',
  'Show HN: Webscope  Easy way for web developers to communicate with Clients',
  'http://webscopeapp.com',
  '3',
  '3',
  'fastbrick',
  '7/28/2016 7:11'],
 ['10872799',
  'Show HN: GeoScreenshot  Easily test Geo-IP based web pages',
  'https://www.geoscreenshot.com/',
  '1',
  '9',
  'kpsychwave',
  '1/9/2016 20:45']]

 let's determine if ask posts or show posts receive more comments on average.

In [10]:
#Total number of comments on Ask posts

total_ask_comments = 0
for row in ask_posts:
    ask_comments = int(row[4])
    sum_comments = total_ask_comments + ask_comments
    total_ask_comments = sum_comments
    
#Average Comments per Post
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [11]:
#Total number of comments in show posts

total_show_comments = 0

for row in show_posts:
    show_comments = int(row[4])
    sum_comments = total_show_comments + show_comments
    total_show_comments = sum_comments
    
#Average Comments Per Post
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


From the above analysis we can see that ask posts get a bit higher comments than the show posts.

Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

* Calculate the number of ask posts created in each hour of the day, along with the number of comments received.
* Calculate the average number of comments ask posts receive by hour created.

In [12]:
#Importing datetime module
import datetime as dt

In [13]:
result_list = []
for row in ask_posts:
    created_at = row[6]
    comments = int(row[4])
    joint = [created_at, comments]
    result_list.append(joint)

 we are creating two dictionaries:

* counts_by_hour: contains the number of ask posts created during each hour of the day.
* comments_by_hour: contains the corresponding number of comments ask posts created at each hour received.

In [14]:
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    time = row[0]
    comment = int(row[1])
    time_datetime = dt.datetime.strptime(time, "%m/%d/%Y %H:%M")
    hour = dt.datetime.strftime(time_datetime, '%H')
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment

In [15]:
counts_by_hour

{'09': 45,
 '13': 85,
 '10': 59,
 '14': 107,
 '16': 108,
 '23': 68,
 '12': 73,
 '17': 100,
 '15': 116,
 '21': 109,
 '20': 80,
 '02': 58,
 '18': 109,
 '03': 54,
 '05': 46,
 '19': 110,
 '01': 60,
 '22': 71,
 '08': 48,
 '04': 47,
 '00': 55,
 '06': 44,
 '07': 34,
 '11': 58}

Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [16]:
avg_by_hour = []

for row in counts_by_hour:
    average = comments_by_hour[row] / counts_by_hour[row]
    avg_by_hour.append([row, average])

In [17]:
#average comments by hours
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

Although we now have the results we need, this format makes it difficult to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [18]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap = [row[1], row[0]]
    swap_avg_by_hour.append(swap)

In [19]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [20]:
#Top 5 Hours for Ask Posts Comments
sorted_swap[:5]

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21']]

In [28]:
for row in sorted_swap[:5]:
    time = row[1]
    time = dt.datetime.strptime(time,'%H')
    time = dt.datetime.strftime(time,'%H:%M:%S')
    average = row[0]
    template = "{time} {average:.2f} average comments per post ."
    summary = template.format(time = time, average = average)
    print(summary)

15:00:00 38.59 average comments per post .
02:00:00 23.81 average comments per post .
20:00:00 21.52 average comments per post .
16:00:00 16.80 average comments per post .
21:00:00 16.01 average comments per post .


### Conclusion
From the above analysis, we can infer that the highest number of comments are seen on the Ask Posts and that too majority are seen during the 15:00 with average comments of 38.59 followed by at 2:00 with 23.81 .
If we take a proper look at our findings, we can see that the top 5 most comments are made during 15:00 - 16:00 and 20:00 - 21:00.
The time of 15:00 - 16:00 usually coincides with the break time and the time evening time after dinner in which people are more likely to be online and comment on posts.